contracts/LockProxySwitcheo.scilla

```diff
@@ -0,0 +1,616 @@
+        event e
+      end
+    | None =>
+      e = IllegalAmount;
+      ThrowError e
+    end
+  end
+end
+
+(* transfer asset from fromAddress to LockProxy contract *)
+procedure TransferToContract(fromAssetHash: ByStr20, amount: Uint128, callAmount: Uint128)
+  (* if people choose to lock native zil, then fromAssetHash should be 0x0000000000000000000000000000000000000000 *)
+  is_asset_zil = builtin eq fromAssetHash zero_address;
+  match is_asset_zil with
+  | True =>
+    accept;
+    is_amount_correct = builtin eq _amount amount;
+    match is_amount_correct with
+    | True =>
+      e = { _eventname: "TransferToContract"; fromAssetHash: fromAssetHash; amount: amount };
+      event e
+    | False =>
+      e = LockAmountMismatch;
+      ThrowError e
+    end
+  | False =>
+    (* no accept here, so even some zils is transfered, there is no harm *)
+    (* see IncreaseAllowance in ZRC2 *)
+    transferZRC2ToContract fromAssetHash _sender _this_address callAmount;
+    e = { _eventname: "TransferToContract"; fromAssetHash: fromAssetHash; amount: callAmount };
+    event e
+  end
+end
+
+procedure CrossChain(toChainId: Uint64, toContract: ByStr, method: ByStr, txData: ByStr)
+  manager_proxy_l <- manager_proxy;
+  msg = {_tag: "CrossChain"; _recipient: manager_proxy_l; _amount: uint128_zero; 
+        toChainId: toChainId; toContract: toContract; method: method; txData: txData };
+  msgs = one_msg msg;
+  send msgs
+end
+
+procedure validateAddrLength(proxyAddr: ByStr)
+  length = builtin strlen proxyAddr;
+  is_addr = builtin eq length uint32_20;
+  match is_addr with
+  | True =>
+  | False =>
+    e = IllegalRegisterAssetArgs;
+    ThrowError e
+  end
+end
+
+(* can be used both for validating targetProxyHash and toAddress and fromAssetHash *)
+procedure ValidataHashLength(hash: ByStr)
+  hash_length = builtin strlen hash;
+  zero_length = builtin eq hash_length uint32_zero;
+  match zero_length with
+  | True =>
+    e = EmptyHashStr;
+    ThrowError e
+  | False =>
+  end
+end
+
+(* Mark an asset as registered by associating it to a specified Switcheo TradeHub proxy and asset hash*)
+procedure MarkAssetAsRegistered(tokenAddr: ByStr20, proxyAddr: ByStr, toAssetHash: ByStr)
+  validateAddrLength proxyAddr;
+  exist <- exists register[tokenAddr];
+  match exist with
+  | True =>
+    e = AssetAlreadyRegistered;
+    ThrowError e
+  | False =>
+    x = serialize_proxy_addr_to_asset_hash proxyAddr toAssetHash;
+    value = builtin keccak256hash x;
+    register[tokenAddr] := value
+  end
+end
+
+(* Validates that an asset registration matches the given params *)
+(* @param tokenAddr: the address of the asset to check *)
+(* @param proxyAddr: the expected proxy address on Swticheo TradeHub *)
+(* @param toAssetHash: the expected asset hash on Swticheo TradeHub *)
+procedure ValidateAssetRegistration(tokenAddr: ByStr20, proxyAddr: ByStr, toAssetHash: ByStr)
+  validateAddrLength proxyAddr;
```

Comment:
need to validate `toAssetHash` ?

Reply:
Already validated in caller transition(unlock and lockProcedure)